interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,373 @@
+//
+// The given notifyPrompt closure should record a notice of type
+// "interfaces-requests-prompt" for the given user with the given
+// promptID as its key.
+func New(notifyPrompt func(userID uint32, promptID string) error) *PromptDB {
+	pdb := PromptDB{
+		perUser:      make(map[uint32]*userPromptDB),
+		notifyPrompt: notifyPrompt,
+	}
+	// Importantly, set maxIDPath before attempting to load max ID
+	pdb.maxIDPath = filepath.Join(dirs.GlobalRootDir, "/tmp/snapd-request-prompt-max-id")
+	// XXX: is /tmp guaranteed to exist, and if not, should we create it here?
+	// Otherwise, should a non-/tmp location be used?
+	err := pdb.loadMaxID()
+	if err != nil {
+		// If cannot read max existing prompt ID, start again from 0.
+		pdb.maxID = 0
+	}
+	return &pdb
+}
+
+// loadMaxID reads the previous ID from the file at maxIDPath and sets maxID.
+//
+// If no file exists at maxIDPath, sets maxID to be 0. If another error occurs,
+// returns it and does not set maxID.
+func (pdb *PromptDB) loadMaxID() error {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+
+	target := pdb.maxIDPath
+	f, err := os.Open(target)
+	if os.IsNotExist(err) {
+		pdb.maxID = 0
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("cannot read maximum prompt ID: %w", err)
+	}
+	defer f.Close()
+
+	idBuf := [16]byte{}
+	_, err = io.ReadFull(f, idBuf[:])
+	if err != nil {
+		return fmt.Errorf("cannot read maximum prompt ID: %w", err)
+	}
+
+	maxID, err := strconv.ParseUint(string(idBuf[:]), 16, 64)
+	if err != nil {
+		return fmt.Errorf("cannot parse maximum prompt ID: %w", err)
+	}
+
+	pdb.maxID = maxID
+	return nil
+}
+
+// nextID advances the internal monotonically-increasing maxID integer.
+//
+// The caller must ensure that the prompt DB mutex is held.
+func (pdb *PromptDB) nextID() string {
+	pdb.maxID++
+	idStr := pdb.maxIDString()
+	osutil.AtomicWriteFile(pdb.maxIDPath, []byte(idStr), 0600, 0)
+	return idStr
+}
+
+// maxIDString returns a 16-character string corresponding to the current maxID.
+// The ID string is the max ID in hexadecimal, padded by leading zeroes.
+func (pdb *PromptDB) maxIDString() string {
+	return fmt.Sprintf("%016X", pdb.maxID)
+}
+
+// AddOrMerge checks if the given prompt contents are identical to an existing
+// prompt and, if so, merges with it by adding the given listenerReq to it.
+// Otherwise, adds a new prompt with the given contents to the prompt DB.
+//
+// If the prompt was merged with an identical existing prompt, returns the
+// existing prompt and true, indicating it was merged. If a new prompt was
+// added, returns the new prompt and false, indicating the prompt was not
+// merged.
+//
+// The caller must ensure that the given permissions are in the order in which
+// they appear in the available permissions list for the given interface.
+func (pdb *PromptDB) AddOrMerge(user uint32, snap string, iface string, path string, permissions []string, listenerReq *listener.Request) (*Prompt, bool) {
```

Comment:
sorry if it's a naiive question, but if one user (A) tries to access files of another user (B), is the `user` here equal to A or B? I suppose it's B, but is the process uid (A) then part of the listener request?

Reply:
The user is the "subject" making the request to access some object (a file, probably). So the user is A here. We don't care about filesystem permissions/ownership here, just apparmor permissions, which can vary depending on the owner of the file, but we let the apparmor profile make those distinctions, and if the request reaches snapd, we no longer care who owns the file.